File: examples/java/saga-loyalty/src/main/java/dev/angzarr/examples/saga/LoyaltySagaService.java
package dev.angzarr.examples.saga;

import com.google.protobuf.Empty;
import io.grpc.stub.StreamObserver;
import dev.angzarr.CommandBook;
import dev.angzarr.EventBook;
import dev.angzarr.SagaGrpc;
import dev.angzarr.SagaResponse;

import java.util.List;

/**
 * gRPC service for loyalty points saga.
 */
public class LoyaltySagaService extends SagaGrpc.SagaImplBase {

    private final LoyaltySaga saga;

    public LoyaltySagaService(LoyaltySaga saga) {
        this.saga = saga;
    }

    @Override
    public void handle(EventBook request, StreamObserver<Empty> responseObserver) {
        saga.processEvents(request);
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void handleSync(EventBook request, StreamObserver<SagaResponse> responseObserver) {
        List<CommandBook> commands = saga.processEvents(request);

        SagaResponse response = SagaResponse.newBuilder()
            .addAllCommands(commands)
            .build();

        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }
}

File: examples/csharp/SagaLoyalty/src/LoyaltySagaService.cs
using Angzarr;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;

namespace Angzarr.Examples.Saga;

public class LoyaltySagaService : Angzarr.Saga.SagaBase
{
    private readonly ILoyaltySaga _saga;

    public LoyaltySagaService(ILoyaltySaga saga)
    {
        _saga = saga;
    }

    public override Task<Empty> Handle(EventBook request, ServerCallContext context)
    {
        _saga.ProcessEvents(request);
        return Task.FromResult(new Empty());
    }

    public override Task<SagaResponse> HandleSync(EventBook request, ServerCallContext context)
    {
        var commands = _saga.ProcessEvents(request);

        var response = new SagaResponse();
        response.Commands.AddRange(commands);

        return Task.FromResult(response);
    }
}
